### ISSUE TITLE
Wrong value of __file__

### ISSUE THREAD

**Issue by joonis** [dashingsoft/pyarmor#12]
The value of __ files __ seems to be hardcoded pointing to the original source file.

Is it possible to fix this?

**Comment by jondy** [comment]
When obfuscate a script, pyarmor first compiles script by calling `Py_CompileString(buf, filename, Py_file_input)`

The value of __file__ is set by `Py_CompileString` implicitly, what's the right value of __file__?

**Comment by jondy** [comment]
Fix it by changing armor code to:
```python
__pyarmor__(__name__, b'...').__file__ = __file__
```

**Comment by joonis** [comment]
But this will set it after the module was executed, isn't it?
So if a module wants to know its location and uses `__file__`, this fails.

Can't you modify `__pyarmor__` to set it before executing the module?
```
__pyarmor__(__name__, __file__, b'...')
```

Alternatively when running Py_CompileString you could set the filename to `<frozen [modname]>`.
Here *modname* should be set to the passed `__name__` param.

**Comment by jondy** [comment]
You're right. I prefer the first way, add extra parameter `__file__`

**Comment by jondy** [comment]
Fixed in v3.7.1, change armor code to
```python
__pyarmor__(__name__, __file__, b'...')
```

**Comment by joonis** [comment]
Now `__file__` is set to the correct path.

But a traceback still contains the wrong path names.

**Comment by jondy** [comment]
I got it. This is because only co_filename of module's code object has been changed, the co_filename of descendants still keep the original value.

**Comment by jondy** [comment]
For filename in traceback, I prefer the second way, that is, set filename to `<frozen [modname]>` when running Py_CompileString. Iterating all the descendants code object to set co_filename in runtime perhaps is not a good choice.

**Comment by joonis** [comment]
Yes, this should be the way to go.

**Comment by joonis** [comment]
Usually the frozen string is without square brackets. Just wanted to indicate the variable for the module name. So it should be:

```
<frozen module>
or
<frozen package.module>
```

**Comment by jondy** [comment]
The dotted module name can not be got directly when running `Py_CompileString`, the alternative way is to take `--src` as the root package. For example,
```bash
python pyarmor.py obfuscate --src=/home/jondy/workspace ...

# /home/jondy/workspace/foo.py -> <frozen foo>
# /home/jondy/workspace/libroot/pkg/foo.py -> <frozen libroot.pkg.foo>  even if "libroot" maybe is not a package
```

**Comment by joonis** [comment]
This shouldn't be a problem. But the square brackets `[]` should be removed.

**Comment by joonis** [comment]
And a package is referenced with `<frozen [__init__]>` at time, instead `<frozen pkgname>`.

**Comment by jondy** [comment]
Fixed in v3.7.3

`$src/__init__.py` will be referenced as `<frozen __init__>`

The sub package `$src/pkgname/__init__.py` will be referenced as `<frozen pkgname.__init__>`

**Comment by joonis** [comment]
Can't you use the passed `__name__` argument for that?

**Comment by jondy** [comment]
There is no `__name__` when obfuscating scripts in building machine, only the full path of source filename.  One solution is to take `__init__.py` as a special case

`__pyarmor__(__name__, __file__, b'...')` run in different time and in different machine. `__file__` and `__name__` is available here, but it need iterate all the code objects to set co_filename as I explains before. For now, only module code object's co_filename is set to `__file__`,  the co_filename of other code objects (functions, methods) is same as build time.

**Comment by joonis** [comment]
Ok, I see. But couldn't you just strip the `.__init__`. part? It looks a bit ugly ;)

**Comment by joonis** [comment]
Hey, I want to get rid of Nuitka for my project. So it's all to my advantage ;)

**Comment by joonis** [comment]
With v3.7.4 I get `<frozen src.pkgname>` and `<frozen pkgname.modname>`.

But it should be: `<frozen pkgname>` and `<frozen pkgname.modname>`

**Comment by joonis** [comment]
If the src folder contains a file `__init__.py`, the src folder itself should be the base package. If there is no such file, the name of the src folder should be skipped.

Example with `__init__.py`:
- `src/`
  - `__init__.py <frozen src>`
  - `subpkg/`
    - `__init__.py <frozen src.subpkg>`
    - `submod.py <frozen src.subpkg.submod>`
  - `mod.py <frozen src.mod>`

Example without `__init__.py`:
- `src/`
  - `subpkg/`
    - `__init__.py <frozen subpkg>`
    - `submod.py <frozen subpkg.submod>`
  - `mod.py <frozen mod>`

**Comment by jondy** [comment]
Fixed in v3.7.5. There is another case
```
src/__init__.py
src/a                                            there is no __init__py
src/a/pkg/__init__.py
src/a/pkg/foo.py                           <frozen pkg.foo>
```

**Comment by joonis** [comment]
I propose that a missing `__init__.py` will always reset package chain.

**Comment by jondy** [comment]
It's perfect. Fixed in v3.8.2
